src/vos/vos_dtx.c

```diff
@@ -1100,3 +1620,257 @@ vos_dtx_mark_sync(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch)
 	vos_obj_release(occ, obj, false);
 	return rc;
 }
+
+/*
+ * \return	+2 if current DTX blob is batched cleanup.
+ *		+1 if current DTX entry has ever been committed.
+ *		Zero if the DTX entry is not committed.
+ *		Negative value if error.
+ */
+static int
+dtx_act_handle_batched_cleanup(struct vos_container *cont,
+			       struct dtx_batched_cleanup_blob *bcb,
+			       struct vos_dtx_act_ent_df *dae_df,
+			       umem_off_t *next_dsb)
+{
+	struct umem_instance		*umm;
+	struct vos_dtx_scm_blob		*dsb;
+	struct vos_dtx_record_df	*rec_df;
+	int				 rc;
+
+	if (daos_is_zero_dti(&dae_df->dae_xid) ||
+	    dae_df->dae_flags & DTX_EF_INVALID)
+		return 1;
+
+	if (dae_df->dae_rec_cnt == 0)
+		return 0;
+
+	umm = vos_cont2umm(cont);
+	rec_df = &dae_df->dae_rec_inline[0];
+	switch (rec_df->dr_type) {
+	case DTX_RT_ILOG:
+		/* XXX: For pure ilog touched DTX case, we do NOT handle it
+		 *	via batched cleanup.
+		 */
+		return 0;
+	case DTX_RT_SVT: {
+		struct vos_irec_df	*svt;
+
+		svt = umem_off2ptr(umm, rec_df->dr_record);
+		if (!dtx_is_null(svt->ir_dtx))
+			return 0;
+		break;
+	}
+	case DTX_RT_EVT: {
+		struct evt_desc		*evt;
+
+		evt = umem_off2ptr(umm, rec_df->dr_record);
+		if (!dtx_is_null(evt->dc_dtx))
+			return 0;
+		break;
+	}
+	default:
+		D_ERROR(DF_UOID" invalid DTX record type %d for DTX "DF_DTI"\n",
+			DP_UOID(dae_df->dae_oid), rec_df->dr_type,
+			DP_DTI(&dae_df->dae_xid));
+		return -DER_IO;
+	}
+
+	rc = vos_tx_begin(umm);
+	if (rc != 0)
+		return rc;
+
+	if (!dtx_is_null(dae_df->dae_rec_off))
+		umem_free(umm, dae_df->dae_rec_off);
+
+	bcb->bcb_dae_count--;
+	dsb = umem_off2ptr(umm, bcb->bcb_dsb_off);
+	if (bcb->bcb_dae_count > 0 || dsb->dsb_index < dsb->dsb_cap) {
+		umem_tx_add_ptr(umm, &dae_df->dae_flags,
+				sizeof(dae_df->dae_flags));
+		dae_df->dae_flags = DTX_EF_INVALID;
+		umem_tx_add_ptr(umm, &dsb->dsb_count, sizeof(dsb->dsb_count));
+		dsb->dsb_count--;
+		rc = 1;
+	} else {
+		dtx_batched_cleanup(cont, bcb, next_dsb);
+		rc = 2;
+	}
+
+	return vos_tx_end(umm, 0);
+}
+
+int
+vos_dtx_act_reindex(struct vos_container *cont)
+{
+	struct umem_instance		*umm = vos_cont2umm(cont);
+	struct vos_cont_df		*cont_df = cont->vc_cont_df;
+	struct vos_dtx_scm_blob		*dsb;
+	umem_off_t			 dsb_off = cont_df->cd_dtx_active_head;
+	d_iov_t				 kiov;
+	d_iov_t				 riov;
+	int				 rc = 0;
+	int				 i;
+
+	while (1) {
```

Comment:
Need to make sure this loop won't hog CPU too long.

Reply:
According to the batched committed mechanism, we will trigger batched commit per 512 DTXs by default, so it is expected that there are very limited (thoughts of level at most) active DTX entries to be handled here. On the other hand, it is mainly read from SCM and build index in DRAM, it is expected to be very quickly.